kroxylicious-sample/src/test/java/io/kroxylicious/sample/SampleFilterIntegrationTest.java

```diff
@@ -212,6 +211,9 @@ void assertConsumerRecordEquals(String value) {
          */
         void close() {
             this.tester.close();
+            this.admin.close();
```

Comment:
Im happy with this as a tactical choice, but longer term? Would a better fix not be to have `test.close()` ensure all the clients it has handed out are closed?

I also wondered about making `FilterIntegrationTest.admin` etc some sort of lazy function

Reply:
I had assumed (incorrectly) that `tester.close()` also closed everything that had been created by the tester (i.e. admin, producer, consumer). Having the tester check what it created and then ensure they're all correctly disposed of on close makes the most sense to me.

Reply:
Sometimes we want the lifespan of the Kafka object (producer, consumer, admin) to extend beyond the lifespan of Kroxylicious instance.  We use this so that we can ensure that clients are capable of reconnecting through a replacement instance.

One way to solve this problem might be to break the association between the lifecycle of the tester the lifecycle of the kroxylicious.  That would retain the flexibility we need to write this sort of test:

```java

try(var tester = kroxyliciousTester();// No longer creates a kroxylicious instance
     var admin = tester.admin()) { 
     try(var kroxy = tester.kroxylicious()) { // Creates a Kroxylicious
        // ...
        //  do something with admin
     } 
     try(var kroxy = tester.kroxylicious()) { // Creates a Kroxylicious
        // ...
        // check admin reconnects
     }
}

It would also mean `tester.close()` could be given the semantics to ensure that everything it created has been closed, which is more obvious for the programmer to reason about.

Reply:
WDYT?  @robobario @gracegrimwood @SamBarker ?

Reply:
Yeah that sounds like a good way to go. I've raised that proposal as #526

Reply:
Does that conflict with your other issue #378, if we can make clients independent of the existence of a Kroxylicious?

Another way would be to make Kroxylicious able to clean up dangling clients on close, and in special cases like the resilience test we construct them in the test using config supplied by the tester.